### DISCUSSION TITLE
OpenWrt .ipk package for a Jamulus headless server

### DISCUSSION THREAD

**Discussion by stevie2sheds** [jamulussoftware/jamulus#2974] [Ideas]
It would be useful to run a headless server on the OpenWrt OS. Also, with a headless server run on OpenWrt there could then be a corresponding LuCI app making headless server management easier.

The Opus libraries and packages are available on OpenWrt, but I can't see any QT easily available to try a straightforward compile as I've previously done on other Linux OS's. It would require an .ipk package compiled for OpenWrt.

**Comment by ann0see** [comment]
Hi,

Thanks for the idea. If getting Qt to work on OpenWRT is difficult, I'd see this as major blocker. What's the current state? Can you compile it from source? If yes, how? Any quirks needed?

The Jamulus team will most likely not maintain/create a packet themselves as that's up to people who are familiar with packaging for specific distros. Do you know anyone who knows about OpenWRT?

Feel free to package it yourself if you know how.

**Reply by stevie2sheds** [reply]
Thanks, for the input. No, I don't know anyone familiar with the OpenWrt distro, but you are correct that there is the right place to start (and end). You're also right in that the lack QT is a major blocker. QT5 is in this feed https://github.com/openwrt/video but I'm new to both this and their toolchain and haven't had success yet even compiling the QT5 modules let alone Jamulus.

FWIW the idea was for a Raspberry Pi 4. The background is that I've recently moved to a full fibre connection and was considering using my Pi, that is currently a headless server, as an OpenWrt router instead. I've got other, lesser, Pis as adblockers etc and OpenWrt works well as an OS for specific purposes, and a Jamulus server would be well suited.

I'll report back if I get anywhere, but for now I'm coming across too many hurdles. The biggest of those is my lack of knowledge!

Many thanks for all your great work btw.

**Reply by ann0see** [reply]
Ok. Although it's probably not perfect, but would a docker container work? I think there's a Jamulus docker container around somewhere.

**Reply by stevie2sheds** [reply]
As I said, my lack of knowledge is my biggest hurdle ;-) Maybe it would work! I'll look into it.

**Comment by hoffie** [comment]
I'm not really into OpenWrt packaging, but in general it should be possible to build Jamulus (maybe even statically) if there's Qt support. The latter might depend on the target platforms. It seems like OpenWrt targets arm, mips(64), x86(64) and ppc. It should certainly be possible to build for arm and x86(64) (not sure about mips and ppc. You can have a look at the existing Linux arm/x86 build logic (it might need some tuning to drop optimizations such as specific arm features or SSE, which may not be available in embedded devices).

This might all be doable on OpenWrt itself, but it may also be easier to perform that on another machine/distro using cross compiling and just packaging the resulting binary.

Building a UI would be a bit more involved, for sure.

**Reply by stevie2sheds** [reply]
> Building a UI would be a bit more involved, for sure. 

Actually that's the part that may be easy and one of the key reasons for the idea. With the LuCI interface you can configure new menus and pages relatively easily. A page to manipulate the startup parameters and toggle recordings is all a headless server would require.

**Comment by ghost** [comment]
A headless build doesn't need Qt, or does it?

**Reply by Rob-NY** [reply]
> A headless build doesn't need Qt, or does it?

Qt is needed for all jamulus implementations; client, server, headless, directory, etc.  Qt provides more than just UI components.  It is the foundational message-based framework for the entire application.

**Reply by jujudusud** [reply]
It is possible to build a "base" version of Qt that can be used on headless cases.

**Comment by stevie2sheds** [comment]
Thanks for the comments. I'm getting there and have the Qt base packages built, albeit at an older version. Now I just need to work on the cross compiling.
